### PR TITLE
Deploy RRGraphBuilder in place of rr_node_indices in routing resource graph builder functions

### PR DESCRIPTION
**Pull request by tangxifan**
This pull request is a follow-up PR on the #1693 

#### Description

- [X] Fully use ``RRGraphBuilder`` when building the lookup in routing resource graph. Eliminate the use of ``rr_node_indice``
- [X] Cut off the dead loop between two header files: ``rr_graph2.h`` and ``rr_graph_storage.h``. Previously, ``rr_graph2.h`` is included in the ``rr_graph_storage.h`` due to a shared data structure ``rr_edge_info_set``. However,  ``rr_graph_storage.h`` should be the primitive header files. Therefore, a new header file ``rr_edge.h`` is created to host the shared data structure ``rr_edge_info_set``.
- [X] Added an API ``resize_nodes()`` to ``RRSpatialLookup``  to pre-allocate memory efficiently
- [X] Added an API ``mirror_nodes()`` to `RRSpatialLookup`` to enable batch copy of indices.

#### Related Issue

This is part of the PR request plan on routing resource graph, being a follow-up PR on the #1693 

#### Motivation and Context

Motivated by @vaughnbetz 's suggestion, the top priority is to deploy the new data structure ``RRGraphBuilder`` in place of the legacy data structure ``rr_node_indices`` in routing resource graph builder functions.
This PR addresses the comment, where we fully remove the use of ``rr_node_indices`` in builder functions.
As a result, ``RRGraphBuilder`` data structure is the only one that builder functions are talking to.

#### How Has This Been Tested?

Should pass all the existing tests, since this PR is about refactoring.

#### Types of changes

- [ ] Bug fix (change which fixes an issue)
- [x] New feature (change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

#### Checklist:

- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly
- [ ] I have added tests to cover my changes
- [ ] All new and existing tests passed

### CODE REVIEW

File: vpr/src/device/rr_spatial_lookup.h
Comment (vaughnbetz) [reply vpr/src/device/rr_spatial_lookup.h:108]: Should also say exactly what this does.  resizes the given 3 dimensions of the RRNodeLookup data structure for the given type.  Any existing data in the lookup for this type is copied over (? I think?)

File: vpr/src/route/rr_graph2.cpp
Comment (vaughnbetz) [inline vpr/src/route/rr_graph2.cpp:940]: We likely should let the .length = 0 case go through eventually, to model muxes.  Could comment that now.

File: vpr/src/route/rr_graph2.cpp
Comment (vaughnbetz) [inline vpr/src/route/rr_graph2.cpp:1098]: Could the whole double-for loop comtaining these calls be moved into RRGraphBuilder as a high-level API?  Call it something like dupLargeBlkLookups().  Then mirror_nodes could be removed or made a (private if possible) helper function.  

Right now the comment for mirror_node describes a good chunk of the body of this for loop, despite the for loop being in the caller.  I think that's a sign we really want a higher-level API.

**Review by vaughnbetz — CHANGES_REQUESTED** [review]

**Comment by vaughnbetz** [comment]
Thanks Xifan.  Looks mostly good, but I have some comments for you to consider.

File: vpr/src/device/rr_spatial_lookup.h
Comment (tangxifan) [reply vpr/src/device/rr_spatial_lookup.h:99]: Thanks for correction. I have adapted the comments about the APIs by following the style in ``vpr_context.h``

Comment (tangxifan) [reply vpr/src/device/rr_spatial_lookup.h:99]: Yes. This function will copy a vector of node indices from a source ``[src_x][src_y][type][side]`` to a destination ``[des_x][des_y][type][side]``. It is required when building the lookup for SOURCE and SINK nodes, which are supposed to be indexable at any location inside a large block ``(x,y)->(x+width, y+height)``.

I agree that we should have a high-level API to do the job, while this low-level API can be a private one. Here are my thoughts:
- We can create a high-level API like ``expand_nodes()`` which will copy the vector from a source coordinate to all the coordinates in a bounding box.
- I also think that it is better to rework the ``find_node()`` API so that we always search the lowest (x,y) corner when dealing with large blocks. But this may require the data structure to be dependent on ``DeviceGrid`` information (it needs to identify if  a grid has height > 1 as well as width >1)

We can keep brainstorming on this front. I have left comments in the header file as a TODO item.

I have also added comments (a quick example) about how to use the API.

File: vpr/src/device/rr_spatial_lookup.h
Comment (tangxifan) [reply vpr/src/device/rr_spatial_lookup.h:108]: Agree. I have reworked the comment to be precise and also comment that any existing data will be maintained without modification.

File: vpr/src/route/rr_graph2.cpp
Comment (tangxifan) [reply vpr/src/route/rr_graph2.cpp:940]: No problem. Added this as a TODO item. I will check details about the ``t_chan_details`` data structure and see what we can do.

File: vpr/src/route/rr_graph2.cpp
Comment (tangxifan) [reply vpr/src/route/rr_graph2.cpp:1098]: Yes. This is what I thought:

- We can create a high-level API like expand_nodes() which will copy the vector from a source coordinate to all the coordinates in a bounding box.

I will keep thinking about what high-level API we need.  I am positive for moving the for-loops into API.

**Comment by tangxifan** [comment]
@vaughnbetz Thanks for the comments. I have addressed all of them and this PR is ready for your review.

Also, we can discuss if more should be added to this PR or leave them in next PRs:
- Use ``find_node()`` API in place of the legacy function ``get_rr_node_index()``. Then we can fully shadow ``rr_node_indices`` and remove it from the ``DeviceContext``
- Get rid of the legacy convention on swapping (x, y) for CHANX nodes when searching its index.
- Develop high-level APIs in place of the ``mirror_nodes()``.

File: vpr/src/route/rr_graph2.cpp
Comment (vaughnbetz) [reply vpr/src/route/rr_graph2.cpp:1099]: Sounds good.  Let's land this and keep iterating on the higher-level api.

**Review by vaughnbetz — APPROVED** [review]
Looks good, thanks! Let's leave further improvements for the follow-on PRs you mention.

**Comment by vaughnbetz** [comment]
Will merge when CI goes green.

**Comment by tangxifan** [comment]
@vaughnbetz  Just the nightly test failed with an error:

```
13:10:43 up  4:06,  0 users,  load average: 3.04, 3.02, 3.00
++ free -h
              total        used        free      shared  buff/cache   available
Mem:           102G        5.5G         60G        8.6M         36G         96G
Swap:          1.0G          0B        1.0G
++ sleep 300
bash: : command not found


[ID: 7125009] Build finished after 14654 secs, exit value: 127


Warning: Permanently added 'localhost' (ED25519) to the list of known hosts.
[13:11:09] Collecting build artifacts from build VM
Build script failed with exit code: 127
```


But this happened after merging the latest master to my branch. I see similar failures in other PRs:  https://github.com/verilog-to-routing/vtr-verilog-to-routing/pull/1728
Not sure why the error happens.

**Comment by vaughnbetz** [comment]
This looks like another spurious kokoro failure; exit code 127 is apparently a command not found one.  I'll merge this and hopefully kokoro recovers soon.

**Comment by tangxifan** [comment]
Thanks @vaughnbetz. I will keep an eye on the regression test and also start working on the next PR.

**Pull request merged**
